src/ucp/rma/amo_sw.c

```diff
@@ -194,39 +194,182 @@ DEFINE_AMO_SW_OP(64)
 DEFINE_AMO_SW_FOP(32)
 DEFINE_AMO_SW_FOP(64)
 
-UCS_PROFILE_FUNC(ucs_status_t, ucp_atomic_req_handler, (arg, data, length, am_flags),
+typedef struct ucp_atomic_flush_ctx {
+    ucp_atomic_req_hdr_t req_hdr;
+    uint64_t             value;
+    ucp_mem_h            memh;
+    ucp_rkey_h           rkey;
+} ucp_atomic_flush_ctx_t;
+
+static void ucp_amo_sw_loopback_send_reply(ucp_worker_h worker,
+                                           ucp_atomic_flush_ctx_t *flush_ctx)
+{
+    ucp_ep_h reply_ep;
+    ucp_request_t *req;
+    ucs_status_t status;
+
+    reply_ep = ucp_worker_get_ep_by_ptr(worker, flush_ctx->req_hdr.req.ep_ptr);
+
+    if (flush_ctx->req_hdr.req.reqptr == 0) {
```

Comment:
most of the code in this function duplicates ucp_atomic_req_handler()
pls extract common code to helper func